### PR TITLE
Fix source select dialog does not add new pages for providers registered after QGIS startup

### PR DESCRIPTION
**Pull request by nyalldawson**
(eg when enabling a new plugin), and also ensure pages are correctly removed when providers are removed

### CODE REVIEW

File: src/gui/qgsdatasourcemanagerdialog.cpp [hunk 1]
```diff
@@ -90,7 +110,8 @@ QgsDataSourceManagerDialog::~QgsDataSourceManagerDialog()
 
 void QgsDataSourceManagerDialog::openPage( const QString &pageName )
 {
-  const int pageIdx = mPageNames.indexOf( pageName );
+  // TODO -- this is actually using provider keys, not provider names!
```
Comment (nirvn) [inline src/gui/qgsdatasourcemanagerdialog.cpp:113]: Still TODO?

**Review by nirvn — APPROVED** [review]

File: src/gui/qgsdatasourcemanagerdialog.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/gui/qgsdatasourcemanagerdialog.cpp:113]: yep -- it's the stable api which is wrong, and I'm not confident to change it.

Comment (nyalldawson) [reply src/gui/qgsdatasourcemanagerdialog.cpp:113]: by that I mean I'm not confidently that if we switch to using keys here internally now we won't break a plugin

Comment (nirvn) [reply src/gui/qgsdatasourcemanagerdialog.cpp:113]: @nyalldawson , fair enough, just wanted to make sure it was a TODO for posterity and not for this PR :)

File: src/gui/qgsdatasourcemanagerdialog.cpp [hunk 2]
```diff
@@ -90,7 +110,8 @@ QgsDataSourceManagerDialog::~QgsDataSourceManagerDialog()
...
+  const int pageIdx = mPageProviderKeys.indexOf( pageName );
```
Comment (agiudiceandrea) [inline src/gui/qgsdatasourcemanagerdialog.cpp:114]: @nyalldawson, the "Vector" page and the "GeoPackage" page have the same "ogr" provider key, so it is not possible to open the GeoPackage page using the openPage function, isn't it?

Comment (nyalldawson) [reply src/gui/qgsdatasourcemanagerdialog.cpp:114]: Actually the "key" referred to here is the QgsSourceSelectProvider providerKey -- it doesn't have to correspond to a QgsDataProvider provider key. So the Geopackage one could (and should) be changed to "gpkg" or similar.

Comment (agiudiceandrea) [reply src/gui/qgsdatasourcemanagerdialog.cpp:114]: > So the Geopackage one could (and should) be changed to "gpkg"

@nyalldawson, I agree. Currently, it seems not possible to add the "Add GeoPackage layer" item to the Layer->Add layer menu: see #51534.

**Pull request merged**